pubsub/google/cloud/pubsub_v1/subscriber/policy/thread.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+from concurrent import futures
+import queue
+import logging
+import threading
+
+import grpc
+
+from google.cloud.pubsub_v1.subscriber import helper_threads
+from google.cloud.pubsub_v1.subscriber.policy import base
+from google.cloud.pubsub_v1.subscriber.message import Message
+
+
+logger = logging.getLogger(__name__)
+
+
+class Policy(base.BasePolicy):
```

Comment:
Basically, the consumer does all RPCs and bookkeeping on the main process. Because of this, it will *always* use threads for its internal concurrency (which the user doesn't actually see). However, internal concurrency strategy does not have any bearing on what is used to execute the callbacks:

1. Threads for bookkeeping and a thread pool executor for callbacks.
2. Threads for bookkeeping and a process pool executor for callbacks.
3. Threads for bookkeeping and a magical executor that sends callbacks to another server.
etc.

As long as the user provides an executor and a queue that can be used to communicate between the main process and the process executing the callbacks, everything can work as expected.